// pareto-core/dist/__internals/sync/assign/decide.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decide = void 0;
exports.ss = ss;
exports.au = au;
const List_1 = require("./literals/List");
/**
 * ss means 'switch state'.
 * used to make the value T the context variable ('$')
 * given a tuple of a string and a value T,
 * the function takes the value T and calls back the callback ($c)
 * notice that the string part is never used
 *
 * example:
 *
 * switch ($.state[0]) {
 *     case 'on':
 *          return ss($.state, ($) => $.value)
 *     case 'off':
 *          return ss($.state, ($) => null)
 *     default: return au($.state[0])
 * }
 */
function ss(option, $c) {
    return $c(option[1]);
}
/**
 * au means 'assert unreachable'. Used in the 'default' clause of switch statements to ensure
 * during compile time that all possible cases have been implemented
 *
 * example:
 *
 * switch (x) {
 *     case '5':
 *         return 5
 *     default: return au(x)
 * }
 *
 * @param _x
 */
function au(_x) {
    throw new Error("unreachable");
}
var decide;
(function (decide) {
    decide.boolean = (boolean_value, if_true, if_false) => boolean_value
        ? if_true()
        : if_false();
    decide.dictionary = (dictionary) => {
        return {
            has_entries: (if_true, if_not_true) => dictionary.__get_number_of_entries() !== 0
                ? if_true(dictionary)
                : if_not_true(),
            has_single_entry: (if_true, if_multiple, if_none) => {
                return decide.list(dictionary.__to_list(($, id) => ({ 'id': id, 'value': $ }))).has_single_item((item) => if_true(item.value, item.id), () => if_multiple(dictionary), if_none);
            }
        };
    };
    decide.list = (list) => {
        return {
            has_first_item: (if_true, if_not_true) => list.__deprecated_get_possible_item_at(0).__decide(($) => if_true($, new List_1.List_Class(list.__get_raw_copy().slice(1))), () => if_not_true()),
            has_items: (if_true, if_not_true) => list.__get_number_of_items() !== 0
                ? if_true(list)
                : if_not_true(),
            has_last_item: (if_true, if_not_true) => list.__deprecated_get_possible_item_at(list.__get_number_of_items() - 1).__decide(($) => if_true($, new List_1.List_Class(list.__get_raw_copy().slice(0, -1))), () => if_not_true()),
            has_match: (test, if_no_match) => {
                const raw = list.__get_raw_copy();
                for (let i = 0; i < raw.length; i++) {
                    const item = raw[i];
                    const result = test(item).__get_raw();
                    if (result !== null) {
                        return result[0];
                    }
                }
                return if_no_match();
            },
            has_single_item: (if_true, if_multiple, if_none) => {
                return list.__get_number_of_items() > 2
                    ? if_multiple(list)
                    : list.__deprecated_get_possible_item_at(0).__decide(($) => if_true($), () => if_none());
            },
        };
    };
    decide.optional = (optional, if_set, if_not_set) => optional.__decide(if_set, if_not_set);
    decide.state = (state, assign) => {
        return assign(state);
    };
    decide.text = (text, assign) => {
        return assign(text);
    };
})(decide || (exports.decide = decide = {}));
