// pareto-core/dist/__internals/async/__query_result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__query_result = __query_result;
const create_refinement_context_1 = require("./create_refinement_context");
class Query_Result_Class {
    constructor(executer) {
        this.executer = executer;
        this.__query_result = null;
    }
    transform_result(transformer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                on_result(transformer($));
            }, on_error);
        });
    }
    query_without_error_transformation(queryer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                queryer($).__extract_data(on_result, on_error);
            }, on_error);
        });
    }
    // query<New_Output, Query_Error>(
    //     queryer: Queryer<New_Output, Query_Error, Output>,
    //     error_transformer: _pi._pi.Transformer<Query_Error, Error>,
    // ): Query_Result<New_Output, Error> {
    //     return new Query_Result_Class<New_Output, Error>((on_result, on_error) => {
    //         this.executer(
    //             ($) => {
    //                 queryer($).__extract_data(
    //                     on_result,
    //                     (queryer_error) => {
    //                         on_error(error_transformer(queryer_error))
    //                     },
    //                 )
    //             },
    //             on_error,
    //         )
    //     })
    // }
    refine_without_error_transformation(callback) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                (0, create_refinement_context_1.create_refinement_context)((abort) => callback($, abort)).__extract_data(on_result, on_error);
            }, on_error);
        });
    }
    rework_error_temp(error_reworker, rework_error_transformer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(on_result, ($) => {
                error_reworker($).__extract_data((new_target_error) => {
                    on_error(new_target_error);
                }, (rework_error) => {
                    on_error(rework_error_transformer(rework_error));
                });
            });
        });
    }
    __extract_data(on_result, on_error) {
        this.executer(on_result, on_error);
    }
}
function __query_result(executer) {
    return new Query_Result_Class(executer);
}
