// pareto-core/dist/__internals/async/command_statement.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.assert = exports.pseudo_query_successfully_executed = exports.if_ = exports.refine_stacked = exports.refine_without_error_transformation = exports.query_stacked = exports.query = exports.fail = exports.nothing = exports.handle_error = exports.dictionaryx = exports.listx = void 0;
const _p_iterate_1 = __importDefault(require("../../_p_iterate"));
const command_promise_1 = require("./command_promise");
const handle_command_block_1 = require("./handle_command_block");
const create_refinement_context_1 = require("./create_refinement_context");
const asynchronous_collection_builder_1 = require("./asynchronous_collection_builder");
const create_asynchronous_processes_monitor_1 = require("./create_asynchronous_processes_monitor");
var listx;
(function (listx) {
    listx.serie = (array) => {
        return (0, command_promise_1.__command_promise)({
            'execute': (on_success, on_error) => {
                (0, _p_iterate_1.default)(array, (iterator) => {
                    const do_next = () => {
                        const next = iterator.look();
                        if (next !== null) {
                            iterator.consume(($) => {
                                $.__start(() => {
                                    do_next();
                                }, on_error);
                            }, {
                                no_more_tokens: () => {
                                    throw new Error("not reachable, just did a look()");
                                }
                            });
                        }
                        else {
                            on_success();
                        }
                    };
                    do_next();
                });
            }
        });
    };
})(listx || (exports.listx = listx = {}));
var dictionaryx;
(function (dictionaryx) {
    dictionaryx.parallel = (dictionary, parametrized_command_block, aggregate_errors) => {
        return (0, command_promise_1.__command_promise)({
            'execute': (on_success, on_error) => {
                const errors_builder = (0, asynchronous_collection_builder_1.create_asynchronous_dictionary_builder)();
                (0, create_asynchronous_processes_monitor_1.create_asynchronous_processes_monitor)((monitor) => {
                    dictionary.__d_map(($, id) => {
                        monitor['report process started']();
                        (0, handle_command_block_1.__handle_command_block)(parametrized_command_block($, id)).__start(() => {
                            monitor['report process finished']();
                        }, (e) => {
                            errors_builder['add entry'](id, e);
                            monitor['report process finished']();
                        });
                    });
                }, () => {
                    const errors = errors_builder['get dictionary']();
                    if (errors.__get_number_of_entries() === 0) {
                        on_success();
                    }
                    else {
                        on_error(aggregate_errors(errors));
                    }
                });
            }
        });
    };
    let deprecated_parallel;
    (function (deprecated_parallel) {
        deprecated_parallel.query = (query_result, parametrized_command_block, aggregate_errors) => {
            return (0, command_promise_1.__command_promise)({
                'execute': (on_success, on_error) => {
                    const errors_builder = (0, asynchronous_collection_builder_1.create_asynchronous_dictionary_builder)();
                    query_result.__extract_data((dictionary) => {
                        (0, create_asynchronous_processes_monitor_1.create_asynchronous_processes_monitor)((monitor) => {
                            dictionary.__d_map(($, id) => {
                                monitor['report process started']();
                                (0, handle_command_block_1.__handle_command_block)(parametrized_command_block($, id)).__start(() => {
                                    monitor['report process finished']();
                                }, (e) => {
                                    errors_builder['add entry'](id, e);
                                    monitor['report process finished']();
                                });
                            });
                        }, () => {
                            const errors = errors_builder['get dictionary']();
                            if (errors.__get_number_of_entries() === 0) {
                                on_success();
                            }
                            else {
                                on_error(aggregate_errors(errors));
                            }
                        });
                    }, on_error);
                }
            });
        };
    })(deprecated_parallel = dictionaryx.deprecated_parallel || (dictionaryx.deprecated_parallel = {}));
})(dictionaryx || (exports.dictionaryx = dictionaryx = {}));
const handle_error = (command_block, parametrized_command_block, assign_target_error) => {
    return (0, command_promise_1.__command_promise)({
        'execute': (on_success, on_error) => {
            (0, handle_command_block_1.__handle_command_block)(command_block).__start(on_success, (e) => {
                (0, handle_command_block_1.__handle_command_block)(parametrized_command_block(e)).__start(() => {
                    on_error(assign_target_error());
                }, on_error);
            });
        }
    });
};
exports.handle_error = handle_error;
const nothing = () => {
    return (0, command_promise_1.__command_promise)({
        'execute': (on_success) => {
            on_success();
        }
    });
};
exports.nothing = nothing;
const fail = (error) => {
    return (0, command_promise_1.__command_promise)({
        'execute': (on_success, on_error) => {
            on_error(error);
        }
    });
};
exports.fail = fail;
const query = (query_result, refine, parametrized_command_block) => {
    return (0, command_promise_1.__command_promise)({
        'execute': (on_success, on_error) => {
            query_result.__extract_data((output) => {
                (0, create_refinement_context_1.create_refinement_context)((abort) => refine(output, abort)).__extract_data(($) => {
                    (0, handle_command_block_1.__handle_command_block)(parametrized_command_block($)).__start(on_success, on_error);
                }, on_error);
            }, on_error);
        }
    });
};
exports.query = query;
const query_stacked = (query_result, parent_data, parametrized_command_block) => {
    return (0, command_promise_1.__command_promise)({
        'execute': (on_success, on_error) => {
            query_result.__extract_data((output) => {
                (0, handle_command_block_1.__handle_command_block)(parametrized_command_block(output, parent_data)).__start(on_success, on_error);
            }, on_error);
        }
    });
};
exports.query_stacked = query_stacked;
const refine_without_error_transformation = (//I doubt that this one is needed. Either parameters are refined or query results.. parameters should already be refined. 
//the only place where parameters are currently refine is in the main functions. can this be solved by adding a refiner to the main function?
callback, parametrized_command_block) => {
    return (0, command_promise_1.__command_promise)({
        'execute': (on_success, on_error) => {
            (0, create_refinement_context_1.create_refinement_context)(callback).__extract_data((output) => {
                (0, handle_command_block_1.__handle_command_block)(parametrized_command_block(output)).__start(on_success, on_error);
            }, on_error);
        }
    });
};
exports.refine_without_error_transformation = refine_without_error_transformation;
const refine_stacked = (callback, parent_data, parametrized_command_block) => {
    return (0, command_promise_1.__command_promise)({
        'execute': (on_success, on_error) => {
            (0, create_refinement_context_1.create_refinement_context)(callback).__extract_data((output) => {
                (0, handle_command_block_1.__handle_command_block)(parametrized_command_block(output, parent_data)).__start(on_success, on_error);
            }, on_error);
        }
    });
};
exports.refine_stacked = refine_stacked;
var if_;
(function (if_) {
    if_.direct = (precondition, command_block, else_command_block) => {
        return (0, command_promise_1.__command_promise)({
            'execute': (on_success, on_error) => {
                if (precondition) {
                    (0, handle_command_block_1.__handle_command_block)(command_block).__start(on_success, on_error);
                }
                else {
                    if (else_command_block !== undefined) {
                        (0, handle_command_block_1.__handle_command_block)(else_command_block).__start(on_success, on_error);
                    }
                    else {
                        on_success();
                    }
                }
            }
        });
    };
    if_.query = (precondition, command_block, else_command_block) => {
        return (0, command_promise_1.__command_promise)({
            'execute': (on_success, on_error) => {
                precondition.__extract_data(($) => {
                    if ($) {
                        (0, handle_command_block_1.__handle_command_block)(command_block).__start(on_success, on_error);
                    }
                    else {
                        if (else_command_block !== undefined) {
                            (0, handle_command_block_1.__handle_command_block)(else_command_block).__start(on_success, on_error);
                        }
                        else {
                            on_success();
                        }
                    }
                }, on_error);
            }
        });
    };
    if_.on_successfully_executed = (command_block, if_true, if_false) => {
        return (0, command_promise_1.__command_promise)({
            'execute': (on_success, on_error) => {
                (0, handle_command_block_1.__handle_command_block)(command_block).__start(() => {
                    (0, handle_command_block_1.__handle_command_block)(if_true()).__start(on_success, on_error);
                }, (e) => {
                    (0, handle_command_block_1.__handle_command_block)(if_false(e)).__start(on_success, on_error);
                });
            }
        });
    };
})(if_ || (exports.if_ = if_ = {}));
const pseudo_query_successfully_executed = (command_block, on_result) => {
    return (0, command_promise_1.__command_promise)({
        'execute': (on_success, on_error) => {
            (0, handle_command_block_1.__handle_command_block)(command_block).__start(() => {
                (0, handle_command_block_1.__handle_command_block)(on_result(true)).__start(on_success, on_error);
            }, (e) => {
                (0, handle_command_block_1.__handle_command_block)(on_result(false)).__start(on_success, on_error);
            });
        }
    });
};
exports.pseudo_query_successfully_executed = pseudo_query_successfully_executed;
var assert;
(function (assert) {
    assert.direct = (assertion, error_if_failed) => {
        return (0, command_promise_1.__command_promise)({
            'execute': (on_success, on_error) => {
                if (!assertion) {
                    on_error(error_if_failed);
                    return;
                }
                on_success();
            }
        });
    };
    assert.query = (assertion, error_if_failed) => {
        return (0, command_promise_1.__command_promise)({
            'execute': (on_success, on_error) => {
                assertion.__extract_data(($) => {
                    if ($) {
                        on_success();
                    }
                    else {
                        on_error(error_if_failed);
                    }
                }, on_error);
            }
        });
    };
})(assert || (exports.assert = assert = {}));
