// pareto-core/dist/select_lookup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stack = exports.cyclic = exports.acyclic = void 0;
var acyclic;
(function (acyclic) {
    acyclic.not_set = () => ({
        get_entry: (id, abort) => abort.no_context_lookup(null),
        __get_entry_raw: (id, abort) => abort.no_context_lookup(null),
    });
    acyclic.from_resolved_dictionary = (dict) => ({
        get_entry: (id, abort) => dict.__get_entry_deprecated(id, {
            no_such_entry: () => abort.no_such_entry(null)
        }),
        __get_entry_raw: (id, abort) => dict.__get_entry_raw(id)
    });
})(acyclic || (exports.acyclic = acyclic = {}));
var cyclic;
(function (cyclic) {
    cyclic.not_set = () => ({
        get_entry: (id, abort) => {
            //return abort['no context lookup']()
            return {
                'get_circular_dependent': () => abort.no_context_lookup(null),
            };
        }
    });
})(cyclic || (exports.cyclic = cyclic = {}));
var stack;
(function (stack_1) {
    stack_1.empty = () => ({
        get_entry: (id, abort) => abort.no_context_lookup(null),
        get_entry_depth(id) {
            return -1;
        },
    });
    stack_1.push = (stack, item) => {
        return ({
            get_entry: (id, abort) => {
                const temp = item.__get_entry_raw(id, abort);
                if (temp === null) {
                    return stack.get_entry(id, abort);
                }
                return temp[0];
            },
            get_entry_depth: (id, abort) => {
                const temp = item.__get_entry_raw(id, abort);
                if (temp === null) {
                    return 1 + stack.get_entry_depth(id, abort);
                }
                return 0;
            }
        });
    };
})(stack || (exports.stack = stack = {}));
