// pareto-core/dist/__internals/sync/assign/construct/dictionary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.from = void 0;
exports.literal = literal;
const Dictionary_1 = require("../literals/Dictionary");
const List_1 = require("../literals/List");
var from;
(function (from) {
    from.dictionary = (dictionary) => {
        return {
            filter: (assign_optional_entry) => {
                const out = {};
                dictionary.__d_map(($, id) => {
                    const result = assign_optional_entry($, id);
                    result.__extract_data((new_value) => {
                        out[id] = new_value;
                    }, () => { });
                });
                return literal(out);
            },
            flatten: (get_child_dictionary, abort) => {
                const out = {};
                dictionary.__d_map(($, id) => {
                    const child_dictionary = get_child_dictionary($);
                    child_dictionary.__d_map(($child, child_id) => {
                        const combined_id = id + child_id;
                        if (out[combined_id] !== undefined) {
                            abort.duplicate_id(combined_id);
                        }
                        out[combined_id] = $child;
                    });
                });
                return literal(out);
            },
            group: (get_id) => {
                const temp = {};
                dictionary.__to_list(($, id) => ({
                    id: id,
                    value: $,
                })).__get_raw_copy().forEach(($) => {
                    const group_id = get_id($.value, $.id);
                    if (temp[group_id] === undefined) {
                        temp[group_id] = {};
                    }
                    temp[group_id][$.id] = $.value;
                });
                return literal(temp).__d_map(($) => literal($));
            },
            join: (other_dictionary, assign_entry) => {
                const out = {};
                dictionary.__d_map(($, id) => {
                    out[id] = assign_entry($, other_dictionary.__get_possible_entry_deprecated(id), id);
                });
                return literal(out);
            },
            map: (assign_entry) => {
                return dictionary.__d_map(assign_entry);
            },
            re_id: (get_id, abort) => {
                const temp = {};
                dictionary.__d_map(($, id) => {
                    const new_id = get_id($, id);
                    if (temp[new_id] !== undefined) {
                        abort.duplicate_id($, new_id);
                    }
                    temp[new_id] = $;
                });
                return literal(temp);
            },
            resolve: (assign_entry) => {
                const source = dictionary;
                const out = {};
                const entries_started = {};
                const cyclic_references = [];
                const inner_resolve = (value, id, stack) => {
                    if (out[id] !== undefined) {
                        // already resolved
                        return;
                    }
                    entries_started[id] = null;
                    out[id] = assign_entry(value, id, {
                        get_entry: (id, abort) => {
                            if (out[id] === undefined) {
                                if (entries_started[id] !== undefined) {
                                    return abort['cycle_detected'](stack.concat([id]));
                                }
                                else {
                                    inner_resolve(source.__get_entry_deprecated(id, {
                                        no_such_entry: () => abort.no_such_entry(null)
                                    }), id, stack.concat([id]));
                                }
                            }
                            // now it must be resolved, otherwise we would have aborted
                            return out[id];
                        },
                        __get_entry_raw: (id, abort) => {
                            const x = source.__get_entry_raw(id);
                            if (x === null) {
                                return null;
                            }
                            else {
                                if (out[id] === undefined) {
                                    if (entries_started[id] !== undefined) {
                                        return abort.cycle_detected(stack.concat([id]));
                                    }
                                    else {
                                        inner_resolve(x[0], id, stack.concat([id]));
                                    }
                                }
                                // now it must be resolved, otherwise we would have aborted
                                return [out[id]];
                            }
                        }
                    }, {
                        get_entry: (id, abort) => {
                            const temp_reference = {
                                'id': id,
                                'value': undefined,
                                'abort': abort,
                            };
                            cyclic_references.push(temp_reference);
                            return {
                                get_circular_dependent: () => {
                                    if (temp_reference.value === undefined) {
                                        return abort.accessing_cyclic_sibling_before_it_is_resolved(null);
                                    }
                                    else {
                                        return temp_reference.value;
                                    }
                                }
                            };
                        }
                    });
                };
                source.__d_map(($, id) => {
                    inner_resolve($, id, [id]);
                });
                cyclic_references.forEach(($) => {
                    const value = out[$.id];
                    if (value === undefined) {
                        $.abort.no_such_entry($.id);
                    }
                    else {
                        $.value = value;
                    }
                });
                return literal(out);
            }
        };
    };
    from.list = (list) => {
        return {
            convert: (get_id, get_value, abort) => {
                const temp = {};
                list.__get_raw_copy().forEach(($) => {
                    const id = get_id($);
                    if (temp[id] !== undefined) {
                        abort.duplicate_id(id);
                    }
                    temp[id] = get_value($);
                });
                return literal(temp);
            },
            group: (get_id) => {
                const temp = {};
                list.__get_raw_copy().forEach(($) => {
                    const id = get_id($);
                    if (temp[id] === undefined) {
                        temp[id] = [];
                    }
                    temp[id].push($);
                });
                return literal(temp).__d_map(($) => new List_1.List_Class($));
            }
        };
    };
})(from || (exports.from = from = {}));
function literal(source) {
    function create_dictionary_as_array(source) {
        const imp = [];
        Object.keys(source).forEach((id) => {
            imp.push({ id: id, value: source[id] });
        });
        return imp;
    }
    return new Dictionary_1.Dictionary_Class(create_dictionary_as_array(source));
}
