// pareto-core/dist/__internals/sync/assign/construct/list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.repeat = exports.nested_literal_new = exports.nested_literal_old = exports.from = void 0;
exports.literal = literal;
const List_1 = require("../literals/List");
var from;
(function (from) {
    from.dictionary = (dictionary) => {
        return {
            convert: (assign_item) => {
                return dictionary.__to_list(assign_item);
            },
            flatten: (assign_item) => {
                const out = [];
                dictionary.__to_list(assign_item).__get_raw_copy().forEach(($) => {
                    const innerList = $;
                    innerList.__get_raw_copy().forEach(($) => {
                        out.push($);
                    });
                });
                return new List_1.List_Class(out);
            },
        };
    };
    from.list = (list) => {
        return {
            filter: (assign_optional_item) => {
                const out = [];
                list.__get_raw_copy().forEach(($) => {
                    const result = assign_optional_item($);
                    result.__extract_data(($) => {
                        out.push($);
                    }, () => { });
                });
                return new List_1.List_Class(out);
            },
            flatten: (assign_list) => {
                const out = [];
                list.__get_raw_copy().forEach(($) => {
                    const innerList = assign_list($);
                    innerList.__get_raw_copy().forEach(($) => {
                        out.push($);
                    });
                });
                return new List_1.List_Class(out);
            },
            join: (other_list, assign_item) => {
                const out = [];
                let index = -1;
                list.__l_map(($) => {
                    index++;
                    out.push(assign_item($, other_list.__deprecated_get_possible_item_at(index)));
                });
                return literal(out);
            },
            map: (assign_item) => {
                return list.__l_map(assign_item);
            },
            reverse: () => {
                return new List_1.List_Class(list.__get_raw_copy().slice().reverse());
            }
        };
    };
})(from || (exports.from = from = {}));
function literal(source) {
    if (!(source instanceof Array)) {
        throw new Error("invalid input in 'list_literal'");
    }
    const data = source.slice(); //create a copy
    /**
     * this is an implementation, not public by design
     * If you feel the need to rename this class, don't rename it to 'Array',
     * it will break the 'instanceOf Array' test
     */
    return new List_1.List_Class(data);
}
const nested_literal_old = (lists) => {
    const out = [];
    lists.forEach(($) => {
        if ($ == undefined) {
            // do nothing
        }
        else if ($ instanceof Array) {
            $.forEach(($) => {
                out.push($);
            });
        }
        else {
            out.push(...$.__get_raw_copy());
        }
    });
    return new List_1.List_Class(out);
};
exports.nested_literal_old = nested_literal_old;
const nested_literal_new = (nested) => {
    const out = [];
    const flatten = (n) => {
        if (n == undefined) {
            // do nothing
        }
        else if (n instanceof Array) {
            n.forEach(($) => {
                flatten($);
            });
        }
        else {
            out.push(...n.__get_raw_copy());
        }
    };
    flatten(nested);
    return new List_1.List_Class(out);
};
exports.nested_literal_new = nested_literal_new;
const repeat = (item, times) => {
    const out = [];
    for (let i = 0; i < times; i++) {
        out.push(item);
    }
    return new List_1.List_Class(out);
};
exports.repeat = repeat;
